weasyl/searchtag.py

```diff
@@ -217,3 +268,170 @@ def tag_history(submitid):
+    if newtags:
+        query.extend(
+            d.engine.execute(
+                "INSERT INTO searchtag (title) SELECT * FROM UNNEST (%(newtags)s) AS title RETURNING tagid, title",
+                newtags=newtags
+            ).fetchall())
+
+    existing_tagids = {t.tagid for t in existing}
+    entered_tagids = {t.tagid for t in query}
+
+    # Assign added and removed
+    added = entered_tagids - existing_tagids
+    removed = existing_tagids - entered_tagids
+
+    if added:
+        if edit_global_blacklist:
+            d.engine.execute("""
+                INSERT INTO searchmapglobalblacklist
+                    SELECT tag, %(uid)s
+                    FROM UNNEST (%(added)s) AS tag
+            """, uid=userid, added=list(added))
+        else:
+            d.engine.execute("""
+                INSERT INTO searchmapuserblacklist
+                SELECT tag, %(uid)s
+                FROM UNNEST (%(added)s) AS tag
+            """, uid=userid, added=list(added))
+
+    if removed:
+        if edit_global_blacklist:
+            d.engine.execute("""
+                DELETE FROM searchmapglobalblacklist
+                WHERE tagid = ANY (%(removed)s)
+            """, removed=list(removed))
+        else:
+            d.engine.execute("""
+                DELETE FROM searchmapuserblacklist
+                WHERE userid = %(uid)s AND tagid = ANY (%(removed)s)
+            """, uid=userid, removed=list(removed))
+
+
+def get_user_searchtag_blacklist(userid):
+    """
+    Retrieves a list of tags on the user searchtag blacklist for friendly display to the user.
+
+    Parameters:
+        userid: The userid of the user requesting the list of tags.
+
+    Returns:
+        A list of blacklisted searchtag titles which were set by ``userid``.
+    """
+    query = d.engine.execute("""
+        SELECT st.title
+        FROM searchmapuserblacklist
+        INNER JOIN searchtag AS st USING (tagid)
+        WHERE userid = %(userid)s
+        ORDER BY st.title ASC
+    """, userid=userid).fetchall()
+    tags = [tag.title for tag in query]
+    return tags
+
+
+def get_global_searchtag_blacklist(userid):
+    """
+    Retrieves a list of tags on the global searchtag blacklist for friendly display to the director.
+
+    Parameters:
+        userid: The userid of the user requesting the list of tags.
+
+    Returns:
+        A list of globally blacklisted searchtag titles and the name of the user which added it.
+    """
+    # Only directors can view the global blacklist; sanity check against the @director_only decorator
+    if userid not in staff.DIRECTORS:
+        raise WeasylError("InsufficientPermissions")
+
+    query = d.engine.execute("""
+        SELECT st.title, lo.login_name
+        FROM searchmapglobalblacklist
+        INNER JOIN searchtag AS st USING (tagid)
+        INNER JOIN login AS lo USING (userid)
+        ORDER BY st.title ASC
+    """).fetchall()
+    return query
+
+
+def query_blacklisted_tags(ownerid, newtagids):
```

Comment:
Could you split this into one function that selects blacklisted tag patterns and another that removes tags matching blacklisted patterns?